src/Microsoft.ML.Data/Scorers/FeatureContributionCalculationTransform.cs

```diff
@@ -0,0 +1,497 @@
+
+                var typeSrc = input.Schema.GetColumnType(colSrc);
+                Func<IRow, int, ValueGetter<VBuffer<float>>> del = GetValueGetter<int>;
+
+                // REVIEW: Assuming Feature contributions will be VBuffer<float>.
+                // For multiclass LR it needs to be(VBuffer<float>[].
+                var meth = del.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(typeSrc.RawType);
+                return (Delegate)meth.Invoke(this, new object[] { input, colSrc });
+            }
+
+            private ReadOnlyMemory<char> GetSlotName(int index, VBuffer<ReadOnlyMemory<char>> slotNames)
+            {
+                var count = slotNames.GetValues().Length;
+                _env.Assert(count > index || count == 0 && slotNames.Length > index);
+                var slotName = slotNames.GetItemOrDefault(index);
+                return slotName.IsEmpty
+                    ? new ReadOnlyMemory<char>($"f{index}".ToCharArray())
+                    : slotName;
+            }
+
+            private ValueGetter<ReadOnlyMemory<char>> GetTextValueGetter<TSrc>(IRow input, int colSrc, VBuffer<ReadOnlyMemory<char>> slotNames)
+            {
+                Contracts.AssertValue(input);
+                Contracts.AssertValue(Predictor);
+
+                var featureGetter = input.GetGetter<TSrc>(colSrc);
+                var map = Predictor.GetFeatureContributionMapper<TSrc, VBuffer<float>>(_topContributionsCount, _bottomContributionsCount, _normalize);
+
+                var features = default(TSrc);
+                var contributions = default(VBuffer<float>);
+                return
+                    (ref ReadOnlyMemory<char> dst) =>
+                    {
+                        featureGetter(ref features);
+                        map(in features, ref contributions);
+                        var indices = new Span<int>();
+                        var values = new Span<float>();
+                        (contributions.IsDense ? Utils.GetIdentityPermutation(contributions.Length).AsSpan() : contributions.GetIndices()).CopyTo(indices);
+                        contributions.GetValues().CopyTo(values);
+                        var count = values.Length;
+                        var sb = new StringBuilder();
+                        GenericSpanSortHelper<int, float>.Sort(indices, values, 0, count);
+                        for (var i = 0; i < count; i++)
+                        {
+                            var val = values[i];
+                            var ind = indices[i];
+                            var name = GetSlotName(ind, slotNames);
+                            sb.AppendFormat("{0}: {1}, ", name, val);
+                        }
+
+                        if (sb.Length > 0)
+                        {
+                            _env.Assert(sb.Length >= 2);
+                            sb.Remove(sb.Length - 2, 2);
+                        }
```

Comment:
a) why we doing this?
b) wouldn't it be easier to have `if (sb.Length>=2)` condition?

Reply:
its not the same, assert here, so if its bigger than 0 then it should be bigger or equal than 2. Dont want to change this logic since this is how it is originally in internal repo

---
In reply to: [236893534](https://github.com/dotnet/machinelearning/pull/1677#discussion_r236893534) [](ancestors = 236893534)